library/src/main/java/com/datatorrent/lib/state/managed/ManagedTimeUnifiedStateImpl.java

```diff
@@ -0,0 +1,243 @@
+ * the same bucket. <br/>
+ * In this implementation multiple partitions can be working with the same time-bucket (since time-bucket is derived
+ * from time). This works because on the disk the time-bucket data is segregated under each operator id.
+ * </li>
+ * <br/>
+ *
+ * <li>While {@link ManagedTimeStateImpl} can support dynamic partitioning by pre-allocating buckets this will not
+ * be able to support dynamic partitioning efficiently.
+ * </li>
+
+ * </ol>
+ */
+public class ManagedTimeUnifiedStateImpl extends AbstractManagedStateImpl implements BucketedState
+{
+  private final transient LinkedBlockingQueue<Long> purgedTimeBuckets = Queues.newLinkedBlockingQueue();
+
+  public ManagedTimeUnifiedStateImpl()
+  {
+    bucketsFileSystem = new TimeUnifiedBucketsFileSystem();
+  }
+
+  @Override
+  public int getNumBuckets()
+  {
+    return timeBucketAssigner.getNumBuckets();
+  }
+
+  @Override
+  public void put(long time, Slice key, Slice value)
+  {
+    long timeBucket = timeBucketAssigner.getTimeBucketFor(time);
+    if (timeBucket == -1) {
+      //time is expired so return null.
+      return;
+    }
+    int bucketIdx = prepareBucket(timeBucket);
+
+    buckets[bucketIdx].put(key, timeBucket, value);
+
+  }
+
+  @SuppressWarnings("SynchronizationOnLocalVariableOrMethodParameter")
+  @Override
+  public Slice getSync(long time, Slice key)
+  {
+    long timeBucket = timeBucketAssigner.getTimeBucketFor(time);
+    if (timeBucket == -1) {
+      //time is expired so return expired slice.
+      return BucketedState.EXPIRED;
+    }
+    int bucketIdx = prepareBucket(timeBucket);
+    Bucket bucket = buckets[bucketIdx];
+
+    synchronized (bucket) {
+      return bucket.get(key, timeBucket, Bucket.ReadSource.ALL);
+    }
+  }
+
+  @SuppressWarnings("SynchronizationOnLocalVariableOrMethodParameter")
+  @Override
+  public Future<Slice> getAsync(long time, Slice key)
+  {
+    long timeBucket = timeBucketAssigner.getTimeBucketFor(time);
+    if (timeBucket == -1) {
+      //time is expired so return null.
+      return Futures.immediateFuture(BucketedState.EXPIRED);
+    }
+    int bucketIdx = prepareBucket(timeBucket);
+    Bucket bucket = buckets[bucketIdx];
+    synchronized (bucket) {
+      Slice cachedVal = buckets[bucketIdx].get(key, timeBucket, Bucket.ReadSource.MEMORY);
+      if (cachedVal != null) {
+        return Futures.immediateFuture(cachedVal);
+      }
+      return readerService.submit(new KeyFetchTask(bucket, key, timeBucket, throwable));
+    }
+  }
+
+  @Override
+  public void endWindow()
+  {
+    super.endWindow();
+    Long purgedTimeBucket;
+
+    //tear down all the purged time buckets
+    while (null != (purgedTimeBucket = purgedTimeBuckets.poll())) {
+      int purgedTimeBucketIdx = getBucketIdx(purgedTimeBucket);
+      if (buckets[purgedTimeBucketIdx] != null && buckets[purgedTimeBucketIdx].getBucketId() == purgedTimeBucket) {
+        buckets[purgedTimeBucketIdx].teardown();
```

Comment:
If there are asynchronous get calls happening when a bucket is purged they may attempt to retrieve data from a closed reader. I would suggest keeping track of the asynchronous get calls and wait for them to finish before purging a time bucket.